# GoodReads Dataset EDA

#### Author: Quinci Birker

### Introduction

### Data Dictionary

| Attributes  | Definition | Completeness |
| ------------- | ------------- | ------------- | 
| bookId  | Book Identifier as in goodreads.com  | 100 |
| title  | Book title | 100 |
| series | Series Name | 45 |
| author | Book's Author | 100 |
| rating | Global goodreads rating | 100 |
| description | Book's description | 97 |
| language | Book's language | 93 |
| isbn | Book's ISBN | 92 |
| genres | Book's genres | 91 |
| characters | Main characters | 26 |
| bookFormat | Type of binding | 97 |
| edition | Type of edition (ex. Anniversary Edition) | 9 |
| pages | Number of pages | 96 |
| publisher | Editorial | 93 |
| publishDate | publication date | 98 |
| firstPublishDate | Publication date of first edition | 59 |
| awards | List of awards | 20 |
| numRatings | Number of total ratings | 100 |
| ratingsByStars | Number of ratings by stars | 97 |
| likedPercent | Derived field, percent of ratings over 2 starts (as in GoodReads) | 99 |
| setting | Story setting | 22 |
| coverImg | URL to cover image | 99 |
| bbeScore | Score in Best Books Ever list | 100 |
| bbeVotes | Number of votes in Best Books Ever list | 100 |
| price | Book's price (extracted from Iberlibro) | 73 |


### Table of Contents

- [Data Cleaning](#data_cleaning)

- [Part 1.2 - Basic Analysis](#q1.2)
---------------------------------------


### Import Dataset and Libraries

In [30]:
# import libraries:
import pandas as pd
import numpy as np
import plotly.express as px

In [31]:
# read data from the CSV file:
raw_df = pd.read_csv('data/books_1.Best_Books_Ever.csv')

In [32]:
# check number of rows and columns:
raw_df.shape
print(f'There are {raw_df.shape[0]} rows and {raw_df.shape[1]} columns in the data')

There are 52478 rows and 25 columns in the data


In [33]:
# check the data types for every column:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  object 
 1   title             52478 non-null  object 
 2   series            23470 non-null  object 
 3   author            52478 non-null  object 
 4   rating            52478 non-null  float64
 5   description       51140 non-null  object 
 6   language          48672 non-null  object 
 7   isbn              52478 non-null  object 
 8   genres            52478 non-null  object 
 9   characters        52478 non-null  object 
 10  bookFormat        51005 non-null  object 
 11  edition           4955 non-null   object 
 12  pages             50131 non-null  object 
 13  publisher         48782 non-null  object 
 14  publishDate       51598 non-null  object 
 15  firstPublishDate  31152 non-null  object 
 16  awards            52478 non-null  object

The majority of these columns are object. Columns that might need to change from object to integer:
- publish date
- first publish date
- price

In [34]:
# sanity check the first five rows:
raw_df.head()

bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                       title                 series  \
0                           The Hunger Games    The Hunger Games #1   
1  Harry Potter and the Order of the Phoenix        Harry Potter #5   
2                      To Kill a Mockingbird  To Kill a Mockingbird   
3                        Pride and Prejudice                    NaN   
4                                   Twilight   The Twilight Saga #1   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   
3  Jane Austen, Anna Quindlen (Introduction)    4.26   
4                            Stephenie Meyer    3.60   

                                         description language           isbn  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English  9780439023481   
1  There is a door at the end of a silent corrido...  English  9780439358071   
2  The unforgettable novel of a childhood in a sl...  English  9999999999999   
3  Alternate cover edition of ISBN 9780679783268S...  English  9999999999999   
4  About three things I was absolutely positive.\...  English  9780316015844   

                                              genres  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          characters  ... firstPublishDate  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  ...              NaN   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  ...         06/21/03   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  ...         07/11/60   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  ...         01/28/13   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  ...         10/05/05   

                                              awards numRatings  \
0  ['Locus Award Nominee for Best Young Adult Boo...    6376780   
1  ['Bram Stoker Award for Works for Young Reader...    2507623   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...    4501075   
3                                                 []    2998241   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...    4964519   

                                      ratingsByStars likedPercent  \
0  ['3444695', '1921313', '745221', '171994', '93...         96.0   
1  ['1593642', '637516', '222366', '39573', '14526']         98.0   
2  ['2363896', '1333153', '573280', '149952', '80...         95.0   
3  ['1617567', '816659', '373311', '113934', '767...         94.0   
4  ['1751460', '1113682', '1008686', '542017', '5...         78.0   

                                             setting  \
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...   
1  ['Hogwarts School of Witchcraft and Wizardry (...   
2               ['Maycomb, Alabama (United States)']   
3  ['United Kingdom', 'Derbyshire, England (Unite...   
4  ['Forks, Washington (United States)', 'Phoenix...   

                                            coverImg  bbeScore bbeVotes  price  
0  https://i.gr-assets.com/images/S/compressed.ph...   2993816    30516   5.09  
1  https://i.gr-assets.com/images/S/compressed.ph...   2632233    26923   7.38  
2  https://i.gr-assets.com/images/S/compressed.ph...   2269402    23328    NaN  
3  https://i.gr-assets.com/images/S/compressed.ph...   1983116    20452    NaN  
4  https://i.gr-assets.com/images/S/compressed.ph...   1459448    14874    2.1  

[5 rows x 25 columns]

In [35]:
# check last 5 rows:
raw_df.tail()

bookId                title  \
52473            11492014-fractured            Fractured   
52474              11836711-anasazi              Anasazi   
52475               10815662-marked               Marked   
52476          11330278-wayward-son          Wayward Son   
52477  10991547-daughter-of-helaman  Daughter of Helaman   

                     series  \
52473            Fateful #2   
52474     Sense of Truth #2   
52475     Soul Guardians #1   
52476                   NaN   
52477  Stripling Warrior #1   

                                                  author  rating  \
52473                   Cheri Schmidt (Goodreads Author)    4.00   
52474                                      Emma Michaels    4.19   
52475                  Kim Richardson (Goodreads Author)    3.70   
52476  Tom Pollack (Goodreads Author), John Loftus (G...    3.85   
52477                    Misty Moncur (Goodreads Author)    4.02   

                                             description language  \
52473  The Fateful Trilogy continues with Fractured. ...  English   
52474  'Anasazi', sequel to 'The Thirteenth Chime' by...  English   
52475  --READERS FAVORITE AWARDS WINNER 2011--Sixteen...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   
52477  Fighting in Helaman's army is Keturah's deepes...  English   

                isbn                                             genres  \
52473  2940012616562  ['Vampires', 'Paranormal', 'Young Adult', 'Rom...   
52474  9999999999999                         ['Mystery', 'Young Adult']   
52475  9781461017097  ['Fantasy', 'Young Adult', 'Paranormal', 'Ange...   
52476  9781450755634  ['Fiction', 'Mystery', 'Historical Fiction', '...   
52477  9781599554976  ['Lds Fiction', 'Historical Fiction', 'Young A...   

      characters  ... firstPublishDate  \
52473         []  ...              NaN   
52474         []  ...  August 3rd 2011   
52475         []  ...  March 15th 2011   
52476         []  ...   April 5th 2011   
52477         []  ...              NaN   

                                        awards numRatings  \
52473                                       []        871   
52474                                       []         37   
52475  ["Readers' Favorite Book Award (2011)"]       6674   
52476                                       []        238   
52477                                       []        246   

                               ratingsByStars likedPercent setting  \
52473       ['311', '310', '197', '42', '11']         94.0      []   
52474             ['16', '14', '5', '2', '0']         95.0      []   
52475  ['2109', '1868', '1660', '647', '390']         84.0      []   
52476           ['77', '78', '59', '19', '5']         90.0      []   
52477          ['106', '73', '42', '17', '8']         90.0      []   

                                                coverImg  bbeScore bbeVotes  \
52473  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52474  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52475  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52476  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52477  https://i.gr-assets.com/images/S/compressed.ph...         0        1   

       price  
52473    NaN  
52474    NaN  
52475   7.37  
52476   2.86  
52477   5.20  

[5 rows x 25 columns]

In [36]:
raw_df.sample(5)

bookId  \
44657          297219.Cloud_Nine   
21692       531197.Lyra_s_Oxford   
43783               6377037-hack   
10013  4371507-the-age-of-wonder   
10205   3313418-extreme-measures   

                                                   title  \
44657                                         Cloud Nine   
21692                                      Lyra's Oxford   
43783                                               Hack   
10013  The Age of Wonder: How the Romantic Generation...   
10205                                   Extreme Measures   

                        series                                       author  \
44657                      NaN               Luanne Rice (Goodreads Author)   
21692  His Dark Materials #3.5  Philip Pullman, John Lawrence (Illustrator)   
43783                      NaN                              Peter Wrenshall   
10013                      NaN                               Richard Holmes   
10205           Mitch Rapp #11                                  Vince Flynn   

       rating                                        description language  \
44657    3.97  Sarah Talbot surely thought she'd never live t...  English   
21692    3.58  Lyra's Oxford begins with Lyra and Pantalaimon...  English   
43783    3.46  Hack, by Peter Wrenshall, is a page-turning yo...  English   
10013    3.96  'The Age of Wonder' is Richard Holmes' first m...  English   
10205    4.30  The latest pulse-pounding thriller by #1 "New ...  English   

                isbn                                             genres  \
44657  9780553589467  ['Romance', 'Fiction', 'Chick Lit', 'Contempor...   
21692  9780375828195  ['Fantasy', 'Young Adult', 'Fiction', 'Short S...   
43783  9781419669972  ['Young Adult', 'Thriller', 'Science Fiction',...   
10013  9780007149520  ['History', 'Science', 'Nonfiction', 'Biograph...   
10205  9781416599395  ['Thriller', 'Fiction', 'Mystery', 'Espionage'...   

                                       characters  ... firstPublishDate  \
44657                                          []  ...         01/01/99   
21692            ['Lyra Belacqua', 'Pantalaimon']  ...         10/30/03   
43783                                          []  ...              NaN   
10013                                          []  ...         10/01/08   
10205  ['Mitch Rapp', 'Mike Nash, Irene Kennedy']  ...              NaN   

                                                  awards numRatings  \
44657                                                 []       4274   
21692                                                 []      21758   
43783                                                 []        120   
10013  ['Royal Society Science Book Prize (2009)', 'A...       8775   
10205                                                 []      48987   

                                  ratingsByStars likedPercent  \
44657      ['1492', '1498', '1030', '188', '66']         94.0   
21692    ['3680', '7577', '8414', '1818', '269']         90.0   
43783              ['23', '41', '28', '24', '4']         77.0   
10013     ['3047', '3226', '1837', '452', '213']         92.0   
10205  ['24683', '16808', '5731', '1117', '648']         96.0   

                   setting                                           coverImg  \
44657                   []  https://i.gr-assets.com/images/S/compressed.ph...   
21692  ['Oxford, England']  https://i.gr-assets.com/images/S/compressed.ph...   
43783                   []  https://i.gr-assets.com/images/S/compressed.ph...   
10013                   []  https://i.gr-assets.com/images/S/compressed.ph...   
10205                   []  https://i.gr-assets.com/images/S/compressed.ph...   

       bbeScore bbeVotes  price  
44657        69        2   5.27  
21692        99        1   5.48  
43783        72        1    NaN  
10013       253        3   2.89  
10205       247        3   5.04  

[5 rows x 25 columns]

Notes from looking at the begining, end, and a random sample of the dataset:

1. Columns to be deleted that are not useful for this project:
    - `bookId` does not seem to be useful for my project. The index will be used instead of this.  
    - `isbn` is a numeric book identifier. I will also delete this since I am using the index to refer to each unique book in the dataset.
    - `coverImg` is a url to the books cover image. For this project, I will not be using this in my modeling.  
2. There are quite a few columns that are missing data. Further analysis will be performed.

3. Columns that stand out for further investigation:
    - `firstPublishDate` ~ missing values and different formats (i.e. 07/29/96, 1989, April 5th 2011)
    - `price` ~ the values don't seem to be accurate. The entire column might need to be deleted

In [37]:
# Check that total index matches total number of rows:
raw_df.index.nunique() == raw_df.shape[0]

True

The total index count is equal to the total number of rows in the dataset. 

In [38]:
# Count the number of missing values for each column:
raw_df.isna().sum()

bookId                  0
title                   0
series              29008
author                  0
rating                  0
description          1338
language             3806
isbn                    0
genres                  0
characters              0
bookFormat           1473
edition             47523
pages                2347
publisher            3696
publishDate           880
firstPublishDate    21326
awards                  0
numRatings              0
ratingsByStars          0
likedPercent          622
setting                 0
coverImg              605
bbeScore                0
bbeVotes                0
price               14365
dtype: int64

In [39]:
# Percentage of missing values for each column:
print(raw_df.isna().sum(axis=0)/raw_df.shape[0])

bookId              0.000000
title               0.000000
series              0.552765
author              0.000000
rating              0.000000
description         0.025496
language            0.072526
isbn                0.000000
genres              0.000000
characters          0.000000
bookFormat          0.028069
edition             0.905579
pages               0.044724
publisher           0.070430
publishDate         0.016769
firstPublishDate    0.406380
awards              0.000000
numRatings          0.000000
ratingsByStars      0.000000
likedPercent        0.011853
setting             0.000000
coverImg            0.011529
bbeScore            0.000000
bbeVotes            0.000000
price               0.273734
dtype: float64


There are 12 columns that have missing values. The column name and rounded percent of missing values in descending order:
- edition: 91%
- series: 55%
- first publish date: 41%
- price 27%
- language: 7%
- publisher: 7%
- pages: 4%
- description: 3%
- book format: 3%
- publish date: 2%
- liked percent: 1%
- cover image: 1%

The first step in the data cleaning process is to delete all the columns that I will not be using for modeling:
- bookid: using dataset index instead
- isbn: using dataset index instead
- cover image: won't be used for this modeling
- edition: over 91% of the values are missing
- first publish date: over 41% missing and my assumption is that publish date will be more relevant than first publish date
- price: the prices do not seem accurate enough to use. Secondly, depending on the date of purchase and retailer, the price of books can very greatly. 
- awards: there is limited values in this dataset
- characters: there is limited values in this dataset
- setting: there is limited values in this dataset

In [49]:
# Check for duplicated columns:
raw_df.T.duplicated().sum()

0

In [58]:
# Checking for duplicate rows: 
raw_df[raw_df.duplicated()]

bookId  \
37431                           8794263-promises-to-keep   
37432                                      1909590.Click   
37433                 23394408-die-unendlichkeit-schl-ft   
37434                                 7544945-death-note   
37435                                  25886017-m-scaras   
37436                      2669775-el-siglo-de-las-luces   
37437                        24633605-always-and-forever   
37438                           237086.Dafne_desvanecida   
37439                                           17786377   
37440    7452583-jag-vill-inte-d-jag-vill-bara-inte-leva   
37441                                           18480662   
37442                        1258333.Cantarella_Volume_1   
37443                                  17999135-exposure   
37444                           220362.Manon_des_sources   
37445                                           23438427   
37446                                  650584.Shadowdale   
37447                           24644867-traiesc-din-nou   
37448                          22587955-count-with-kenny   
37449             8438589-the-creatures-that-time-forgot   
37450                          17337543-the-game-changer   
37451                       22372599-the-seahorse-legacy   
37452                    9807272-big-nate-boredom-buster   
37453                          635270.The_Planet_Pirates   
37454              23590446-how-to-fight-a-dragon-s-fury   
37455                                 60614.Diamond_Dogs   
37456                             16058645-the-bookstore   
37457          15733451-chairul-tanjung-si-anak-singkong   
37458                                           25454580   
37459                              3486235-night-letters   
37460                         11737313-three-times-lucky   
37461                              13614864-know-not-why   
37462                              21900147-gone-too-far   
37463                      15741217-writing-scary-scenes   
37464  6665889-the-science-fiction-hall-of-fame-volum...   
37465                                2081377.Ghosthunter   
37466                                  3973924-bal-masqu   
37467                        18849994-beautifully-broken   
37468  8538135-catfancy-s-naughty-no-more-change-unwa...   
37469                 18813012-the-rocker-that-holds-her   
37470                                           17406982   
37471                                 21254063-sorceress   
37472                                25590067-conviction   
37473                           383514.Her_Royal_Spyness   
37474              2074104.The_Legend_of_Colton_H_Bryant   
37475                      20262067-le-festin-du-serpent   
37476                                   24903989-widz-ci   
37477                         975953.Time_of_the_Dragons   
37478                                    23519340-silber   
37479                              29742514-mays-landing   
37480                                26845691-beatlebone   
37481                22916606-gli-enigmi-dei-vedovi-neri   
37482                                         13571757-1   
37483    1481056.Gifts_and_Ministries_of_the_Holy_Spirit   
37484                        27074409-deadly-class-vol-1   

                                                   title  \
37431                                   Promises to Keep   
37432                                              Click   
37433                          Die Unendlichkeit schläft   
37434                  Death Note: Black Edition, Vol. 2   
37435                                           Máscaras   
37436                              El siglo de las luces   
37437                                 Always and Forever   
37438                                  Dafne desvanecida   
37439                                       فضولية العلم   
37440          Jag vill inte dö, jag vill bara inte leva   
37441                     لا سكاكين في مطابخ هذه المدينة   
37442                        

These duplicated rows have all the same bbeScore and bbeVotes, yet all the book titles are unique, meaning that these rows will not be dropped.

<a id = 'data_cleaning'></a>
### Data Cleaning

First step in data cleaning is dropping the columns that will not be used in the analysis/modeling.

In [63]:
# Drop specified column and check that changes have been made to raw_df:

raw_df = raw_df.drop(['isbn', 'coverImg', 'edition', 'firstPublishDate', 'price', 'awards', 'setting', 'characters'], axis=1)
raw_df.head()

KeyError: "['isbn', 'coverImg', 'edition', 'firstPublishDate', 'price', 'awards', 'setting', 'characters'] not found in axis"

For the publish date, the first 30,000 books in the dataset are formatted in mm/dd/yyyy while the last 22,478 books are formated in Month Day Year. I will reformate all the dates to be in mm/dd/yyyy.

In [71]:
# Change publish date to datetime format:
raw_df['publishDate'] = pd.to_datetime(raw_df['publishDate'], errors='coerce') # errors='coerce' used if value cannot be converted to datetime format
# Reformat publish date to mm/dd/yyyy format:
raw_df['publishDate'] = raw_df['publishDate'].dt.strftime('%m/%d/%Y')

Sanity check that this change was made to the dataset:

In [69]:
raw_df.sample(5)

bookId                                title  \
45664      919341.Above_San_Diego                      Above San Diego   
7641     10002296-wildflower-hill                      Wildflower Hill   
6052   17204984-between-the-lives                    Between the Lives   
13263           15790912-new-moon  New Moon: The Graphic Novel, Vol. 2   
22632             18460726-strain                               Strain   

                               series  \
45664                             NaN   
7641                              NaN   
6052                              NaN   
13263  Twilight: The Graphic Novel #4   
22632                       Strain #1   

                                                  author  rating  \
45664  Robert W. Cameron (Photographs), Neil Morgan (...    4.00   
7641                                   Kimberley Freeman    4.09   
6052              Jessica Shirvington (Goodreads Author)    4.21   
13263  Young Kim (Adaptation), Stephenie Meyer (Creator)    4.33   
22632               Amelia C. Gormley (Goodreads Author)    3.87   

                                             description language  \
45664  Above San Diego. This is the anchor of the wes...  English   
7641   Forced to take her life in a new direction whe...  English   
6052   Sabine isn't like anyone else. For as long as ...  English   
13263  Bella and Edward find themselves facing new ob...  English   
22632  In a world with little hope and no rules, the ...  English   

                                                  genres bookFormat pages  \
45664                                                 []  Hardcover   160   
7641   ['Historical Fiction', 'Fiction', 'Romance', '...  Paperback   524   
6052   ['Young Adult', 'Paranormal', 'Romance', 'Fant...  Paperback   336   
13263  ['Graphic Novels', 'Vampires', 'Young Adult', ...  Hardcover   160   
22632  ['M M Romance', 'BDSM', 'Science Fiction', 'Dy...      ebook   375   

                     publisher publishDate  numRatings  \
45664            Cameron Books  08/01/1990           9   
7641                Touchstone  08/23/2011       12163   
6052   HarperCollins Australia  05/01/2013        4906   
13263                Yen Press  07/21/2020        1164   
22632       Riptide Publishing  02/17/2014        1777   

                               ratingsByStars  likedPercent  bbeScore  \
45664               ['2', '5', '2', '0', '0']         100.0        65   
7641   ['4268', '5271', '2149', '362', '113']          96.0       348   
6052     ['2314', '1629', '696', '207', '60']          95.0       465   
13263       ['733', '221', '121', '43', '46']          92.0       186   
22632      ['571', '664', '353', '124', '65']          89.0        99   

       bbeVotes  
45664         1  
7641          4  
6052          5  
13263         2  
22632         1

Next step is to go through the columns with missing values. First, looking at series which is missing 55% of it's values.

In [42]:
raw_df.loc[raw_df['series'].isna()]

bookId  \
3                1885.Pride_and_Prejudice   
5                    19063.The_Book_Thief   
6                      170448.Animal_Farm   
9                18405.Gone_with_the_Wind   
10        11870085-the-fault-in-our-stars   
...                                   ...   
52463  25876358-the-natural-way-of-things   
52464                  36374396-algedonic   
52469               270435.Heal_Your_Body   
52470          11115191-attracted-to-fire   
52476                11330278-wayward-son   

                                                   title series  \
3                                    Pride and Prejudice    NaN   
5                                         The Book Thief    NaN   
6                                            Animal Farm    NaN   
9                                     Gone with the Wind    NaN   
10                                The Fault in Our Stars    NaN   
...                                                  ...    ...   
52463                          The Natural Way of Things    NaN   
52464                                          Algedonic    NaN   
52469  Heal Your Body: The Mental Causes for Physical...    NaN   
52470                                  Attracted to Fire    NaN   
52476                                        Wayward Son    NaN   

                                                  author  rating  \
3              Jane Austen, Anna Quindlen (Introduction)    4.26   
5                        Markus Zusak (Goodreads Author)    4.37   
6      George Orwell, Russell Baker (Preface), C.M. W...    3.95   
9                                      Margaret Mitchell    4.30   
10                         John Green (Goodreads Author)    4.21   
...                                                  ...     ...   
52463                                     Charlotte Wood    3.53   
52464                        R.H. Sin (Goodreads Author)    3.71   
52469                                      Louise L. Hay    4.36   
52470                     DiAnn Mills (Goodreads Author)    4.14   
52476  Tom Pollack (Goodreads Author), John Loftus (G...    3.85   

                                             description language  \
3      Alternate cover edition of ISBN 9780679783268S...  English   
5      Librarian's note: An alternate cover edition c...  English   
6      Librarian's note: There is an Alternate Cover ...  English   
9      Scarlett O'Hara, the beautiful, spoiled daught...  English   
10     Despite the tumor-shrinking medical miracle th...  English   
...                                                  ...      ...   
52463  Two women awaken from a drugged sleep to find ...  English   
52464  Bestselling poet r.h. Sin, author of the Whisk...      NaN   
52469  Heal Your Body is a fresh and easy step-by-ste...  English   
52470  Special Agent Meghan Connors' dream of one day...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   

                                                  genres  \
3      ['Classics', 'Fiction', 'Romance', 'Historical...   
5      ['Historical Fiction', 'Fiction', 'Young Adult...   
6      ['Classics', 'Fiction', 'Dystopia', 'Fantasy',...   
9      ['Classics', 'Historical Fiction', 'Fiction', ...   
10     ['Young Adult', 'Romance', 'Fiction', 'Contemp...   
...                                                  ...   
52463  ['Fiction', 'Dystopia', 'Australia', 'Feminism...   
52464    ['Poetry', 'Nonfiction', 'Romance', 'Feminism']   
52469  ['Self Help', 'Health', 'Nonfiction', 'Spiritu...   
52470  ['Christian Fiction', 'Christian', 'Suspense',...   
52476  ['Fiction', 'Mystery', 'Historical Fiction', '...   

                  bookFormat pages                  publisher  \
3                  Paperback   279             Modern Library   
5                  Hardcover   552            Alfred A. Knopf   
6      Mass Market Paperback   141            Signet Classics   
9      Mass Market Paperback  1037               Warner Books 

# COME BACK TO THIS ^^

Language column has 3,806 (7 percent) missing values. Further investigation will happen here:

In [72]:
# Look at the rows with language column missing:
raw_df.loc[raw_df['language'].isna()]

bookId  \
503                34521870-mistress-suffragette   
570                  36236125-invisible-monsters   
645               38311414-house-of-sand-and-fog   
703                       41423092-the-awakening   
751    7770.One_Fish_Two_Fish_Red_Fish_Blue_Fish   
...                                          ...   
52411                        21691391-pixie-dust   
52416                       23014216-when-i-fall   
52429   20665259-did-god-really-command-genocide   
52454                     13254398-au-petit-poil   
52464                         36374396-algedonic   

                                                   title  \
503                                 Mistress Suffragette   
570                                   Invisible Monsters   
645                                House of Sand and Fog   
703                        The Awakening: Fate in Motion   
751              One Fish, Two Fish, Red Fish, Blue Fish   
...                                                  ...   
52411                                         Pixie Dust   
52416                                        When I Fall   
52429  Did God Really Command Genocide?: Coming to Te...   
52454                                      Au Petit Poil   
52464                                          Algedonic   

                         series                               author  rating  \
503                         NaN      Diana Forbes (Goodreads Author)    4.34   
570                         NaN   Chuck Palahniuk (Goodreads Author)    3.98   
645                         NaN                      Andre Dubus III    3.85   
703                         NaN  Suzanne Boisvert (Goodreads Author)    4.31   
751                         NaN                   Dr. Seuss (Reader)    4.13   
...                         ...                                  ...     ...   
52411  Pixie Dust Chronicles #1         Laura Lee (Goodreads Author)    3.79   
52416         Alabama Summer #3        J. Daniels (Goodreads Author)    4.30   
52429                       NaN           Paul Copan, Matt Flannagan    4.18   
52454          Cool and Lam #26                            A.A. Fair    3.89   
52464                       NaN          R.H. Sin (Goodreads Author)    3.71   

                                             description language  \
503    \n  A young woman without prospects at a ball ...      NaN   
570    She's a catwalk model who has everything: a bo...      NaN   
645    In this “page-turner with a beating heart” (Bo...      NaN   
703    Exiled from Earth thousands of years ago, Sar ...      NaN   
751    One Fish Two Fish Red Fish Blue Fish is a 1960...      NaN   
...                                                  ...      ...   
52411  *A lonesome fairy with no clue how to wield he...      NaN   
52416  From New York Times bestselling author, J. Dan...      NaN   
52429  A common objection to belief in the God of the...      NaN   
52454                                                NaN      NaN   
52464  Bestselling poet r.h. Sin, author of the Whisk...      NaN   

                                                  genres      bookFormat  \
503    ['Fiction', 'Novels', 'Historical Fiction', 'D...  Kindle Edition   
570    ['Fiction', 'Contemporary', 'Thriller', 'Myste...       Paperback   
645    ['Fiction', 'Contemporary', 'Literary Fiction'...       Paperback   
703    ['Contemporary', 'Drama', 'Book Club', 'Fictio...  Kindle Edition   
751    ['Childrens', 'Picture Books', 'Fiction', 'Cla...       Hardcover   
...                                                  ...             ...   
52411  ['Fantasy', 'Paranormal', 'Paranormal Romance'...           ebook   
52416  ['Romance', 'New Adult', 'Contemporary Romance...             NaN   
52429  ['Theology', 'Christian', 'Religion', 'Christi...       Paperback   
52454                ['Mystery', 'Fiction', 'Detective']             NaN   
52464    ['Poetry', 'Nonfiction', 'Romance', 'Feminism']       Paperba

The pages column is missing around 4% of it's data. Further investigation...

In [73]:
# Look at the rows with pages column missing:
raw_df.loc[raw_df['pages'].isna()]

bookId  \
669                                 30346601-brainwalker   
1062                               44575575-pleasant-day   
1428                             318525.Red_Storm_Rising   
1483                              26030383-beg-for-mercy   
1543                       19181419-a-bird-without-wings   
...                                                  ...   
52295         53946399-launching-your-real-estate-career   
52352                               971322.The_Wanderers   
52359                           18130046-ender-in-flight   
52407  2823038-forever-young-forever-free-forever-you...   
52411                                21691391-pixie-dust   

                                                   title  \
669                                          Brainwalker   
1062                                        Pleasant Day   
1428                                    Red Storm Rising   
1483                                       Beg For Mercy   
1543                                A Bird Without Wings   
...                                                  ...   
52295  Launching Your Real Estate Career: The Definit...   
52352                                      The Wanderers   
52359                                    Ender in Flight   
52407  Forever Young, Forever Free Forever Young, For...   
52411                                         Pixie Dust   

                           series  \
669                           NaN   
1062                          NaN   
1428                          NaN   
1483                     Mercy #3   
1543                          NaN   
...                           ...   
52295                         NaN   
52352                         NaN   
52359  Ender's Saga short stories   
52407                         NaN   
52411    Pixie Dust Chronicles #1   

                                                  author  rating  \
669    Robyn Mundell (Goodreads Author), Stephan Laca...    4.35   
1062                   Vera Jane Cook (Goodreads Author)    4.33   
1428                                          Tom Clancy    4.17   
1483                      Lucian Bane (Goodreads Author)    4.29   
1543                   Roberta Pearce (Goodreads Author)    4.17   
...                                                  ...     ...   
52295                                       Eric Kistner    5.00   
52352                                      Richard Price    3.90   
52359                                   Orson Scott Card    4.35   
52407                                       Hettie Jones    3.67   
52411                       Laura Lee (Goodreads Author)    3.79   

                                             description language  \
669    One teen’s incredible journey may just blow hi...  English   
1062   In the town of Hollow Creek, South Carolina tw...  English   
1428   "Allah!"With that shrill cry, three Muslim ter...  English   
1483   The fight is on in this installment, Mercy is ...  English   
1543   After an impoverished and indigent childhood, ...  English   
...                                                  ...      ...   
52295                                                NaN      NaN   
52352  The Wanderers, a teen-age gang in the Bronx of...  English   
52359  "Ender in Flight" is a story by Orson Scott Ca...  English   
52407                                                NaN      NaN   
52411  *A lonesome fairy with no clue how to wield he...      NaN   

                                                  genres  \
669    ['Young Adult', 'Adventure', 'Fantasy', 'Middl...   
1062   ['Contemporary', 'Drama', 'Book Club', 'Fictio...   
1428   ['Fiction', 'Thriller', 'Military Fiction', 'W...   
1483                                                  []   
1543   ['Romance', 'Contemporary Romance', 'Contempor...   
...                                                  ...   
52295                                                 []   
52352  ['Fiction', 'Crime', 'New York